runtime/src/bank.rs

```diff
@@ -898,77 +899,135 @@ impl Bank {
             let inflation = self.inflation.read().unwrap();
 
             (*inflation).validator(year) * self.capitalization() as f64 * period
-        };
+        } as u64;
+
+        let vote_balance_and_staked = self.stakes.read().unwrap().vote_balance_and_staked();
+
+        let validator_point_value = self.pay_validator_rewards(validator_rewards);
 
-        let validator_points = self.stakes.write().unwrap().claim_points();
-        let validator_point_value =
-            self.check_point_value(validator_rewards / validator_points as f64);
+        // this sysvar could be retired...
         self.update_sysvar_account(&sysvar::rewards::id(), |account| {
             sysvar::rewards::create_account(
                 self.inherit_sysvar_account_balance(account),
                 validator_point_value,
             )
         });
 
-        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+        let validator_rewards_paid =
+            self.stakes.read().unwrap().vote_balance_and_staked() - vote_balance_and_staked;
+        if let Some(rewards) = self.rewards.as_ref() {
+            assert_eq!(
+                validator_rewards_paid,
+                u64::try_from(rewards.iter().map(|(_pubkey, reward)| reward).sum::<i64>()).unwrap()
+            );
+        }
+
+        // verify that we didn't pay any more than we expected to
+        assert!(validator_rewards >= validator_rewards_paid);
 
         self.capitalization
-            .fetch_add(validator_rewards as u64, Ordering::Relaxed);
+            .fetch_add(validator_rewards_paid, Ordering::Relaxed);
     }
 
-    /// iterate over all stakes, redeem vote credits for each stake we can
-    ///   successfully load and parse, return total payout
-    fn pay_validator_rewards(&mut self, point_value: f64) -> u64 {
-        let stake_history = self.stakes.read().unwrap().history().clone();
-        let mut validator_rewards = HashMap::new();
-
-        let total_validator_rewards = self
+    /// map stake delegations into resolved pubkeys
+    ///  returns a vector (has to be copied) of loaded
+    ///   ( staker pubkey voter pubkey )
+    ///
+    fn stake_delegation_pubkeys(&self) -> Vec<(Pubkey, Pubkey)> {
```

Comment:
(repost of https://github.com/solana-labs/solana/commit/8472fc3f0825c1c2d4a68050bd667095d517081e#r40694986)

> nice catch.
> 
> this could also return a hashmap ordered by the voter pubkey, then we'd re-organize the loop to store the vote accounts just once

Reply:
Commenting on https://github.com/solana-labs/solana/blob/0218d32705392151c5e8e4e043d7709aefbcef72/runtime/src/bank.rs#L941

(repost of https://github.com/solana-labs/solana/pull/10914#discussion_r457000764):

> @ryoqun I think this approach still has merit, if the `store_account()` phase is moved until after all the stakes have been processed, or if the vote_account is modified while in the vector

Reply:
@rwalker-com Yeah, I generally agree that your suggestion is more cleaner than my hacky approach. Reverted mine (https://github.com/solana-labs/solana/pull/10914/commits/8608e4eb19f01bb79aa893355c8998d33fcb100a) and I think https://github.com/solana-labs/solana/pull/10914/commits/ef5f158ea47a8dd7bc8f63db49eeb74dc4500ceb is what you meant.